go/dbagent/dbagent.go

```diff
@@ -0,0 +1,360 @@
+	MySQLDatadirPath     string
+	MySQLBackupdirPath   string
+	IsSlave              bool
+	IsMaster             bool
+	IsBinlogEnabled      bool
+	HasActiveConnections bool
+}
+
+func OpenConnection() (*sql.DB, error) {
+	mysqlURI := fmt.Sprintf("%s:%s@tcp(%s:%d)/%s?interpolateParams=true",
+		config.Config.MySQLTopologyUser,
+		config.Config.MySQLTopologyPassword,
+		"localhost",
+		config.Config.MySQLPort,
+		"mysql",
+	)
+	db, _, err := sqlutils.GetDB(mysqlURI)
+	db.SetMaxIdleConns(0)
+	if err != nil {
+		return nil, log.Errore(err)
+	}
+	err = db.Ping()
+	if err != nil {
+		return nil, log.Errore(err)
+	}
+	return db, err
+}
+
+func QueryData(query string, argsArray []interface{}, on_row func(sqlutils.RowMap) error) error {
+	db, err := OpenConnection()
+	if err != nil {
+		return log.Errore(err)
+	}
+	return log.Criticale(sqlutils.QueryRowsMap(db, query, on_row, argsArray...))
+}
+
+func ExecuteQuery(query string, args ...interface{}) (sql.Result, error) {
+	var err error
+	db, err := OpenConnection()
+	if err != nil {
+		return nil, err
+	}
+	res, err := sqlutils.ExecNoPrepare(db, query, args...)
+	return res, err
+}
+
+func GetMySQLVersion() (Version string, err error) {
+	query := `SELECT @@version;`
+	err = QueryData(query, sqlutils.Args(), func(m sqlutils.RowMap) error {
+		Version = m.GetString("@@version")
+		return nil
+	})
+	if err == nil {
+		Version = Version[0:strings.LastIndex(Version, ".")]
+	}
+	if err != nil {
+		log.Errore(err)
+	}
+	return Version, err
+}
+
+func GetMySQLSql_mode() (sqlMode string, err error) {
+	query := `SELECT @@sql_mode;`
+	err = QueryData(query, sqlutils.Args(), func(m sqlutils.RowMap) error {
+		sqlMode = m.GetString("@@sql_mode")
+		return nil
+	})
+	if err != nil {
+		log.Errore(err)
+	}
+	return sqlMode, err
+}
+
+func SetMySQLSql_mode(sqlMode string) error {
+	query := fmt.Sprintf("SET GLOBAL SQL_MODE='%s';", sqlMode)
+	_, err := ExecuteQuery(query)
+	return err
+}
+
+func isBinlogEnabled() (IsBinlogEnabled bool, err error) {
+	query := `SHOW VARIABLES LIKE 'log_bin';`
+	err = QueryData(query, sqlutils.Args(), func(m sqlutils.RowMap) error {
+		IsBinlogEnabled = (m.GetString("Value") == "ON")
+		return nil
+	})
+	if err != nil {
+		log.Errore(err)
+	}
+	return IsBinlogEnabled, err
+}
+
+func isSlave() (IsSlave bool, err error) {
+	query := `SHOW SLAVE STATUS;`
+	err = QueryData(query, sqlutils.Args(), func(m sqlutils.RowMap) error {
+		ioThreadRunning := (m.GetString("Slave_IO_Running") == "Yes")
+		sqlThreadRunning := (m.GetString("Slave_SQL_Running") == "Yes")
+		IsSlave = ioThreadRunning && sqlThreadRunning
+		return nil
+	})
+	if err != nil {
+		log.Errore(err)
+	}
+	return IsSlave, err
+}
+
+func isMaster() (IsMaster bool, err error) {
+	query := `SHOW SLAVE HOSTS;`
+	err = QueryData(query, sqlutils.Args(), func(m sqlutils.RowMap) error {
+		IsMaster = (len(m.GetString("Slave_UUID")) != 0)
+		return nil
+
+	})
+	if err != nil {
+		log.Errore(err)
+	}
+	return IsMaster, err
```

Comment:
Incorrect. This could be an intermediate master.
Better to run `SHOW SLAVE STATUS` and see that there's no parent. Though that would fail with master-master.

Or... ask `orchestrator`.

Reply:
Agree. Better implement all these checks on orchestrator side before starting provisioning. Removed in b73ea71